### PR TITLE
Add a GitHub Action to automatically compress image files

### PR DESCRIPTION
**Pull request by sgibson91**
### Summary

Image files will be the biggest cause of "size bloat" for this repository. We already check for particularly large files (of any kind) that are added to the repo, but another solution is to automatically compress images when they are added (I suspect not all contributors know how to do this or why it could be an issue).

This PR adds a new GitHub Action workflow that will automatically compress images changed in a Pull Request using the [Crush.pics API](https://crush.pics/). NOTE: It will not touch images already on `main`, only new images being added via a PR. I have created a free account with crush.pics and added an API token to the repository secrets.

### List of changes proposed in this PR (pull-request)

* `.github/workflows/auto-compress-images.yaml`

### What should a reviewer concentrate their feedback on?

- [ ] Everything looks ok?

### Acknowledging contributors

- [x] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/main/README.md#contributors) in the README file:

### CODE REVIEW

**Comment by malvikasharan** [comment]
@sgibson91 this looks great! I haven't tested it, which I will do this evening.
Thank you so much for the prompt PR <3

**Comment by sgibson91** [comment]
I'm marking this as draft again as I realised, in it's current state, that this PR will never work. Thank you to Arron for alerting me to this fact.

Basically, we'll run into the "secrets are not available to forks" problem (again 😫) and so I'll have to do some [`test-this-pr`](https://blog.jupyter.org/how-i-automated-authorised-cloud-deployments-from-pull-requests-with-github-actions-13f890538e32)-like magic to sort that out

**Comment by arronlacey** [comment]
> I'm marking this as draft again as I realised, in it's current state, that this PR will never work. Thank you to Arron for alerting me to this fact.
> 
> Basically, we'll run into the "secrets are not available to forks" problem (again 😫) and so I'll have to do some [`test-this-pr`](https://blog.jupyter.org/how-i-automated-authorised-cloud-deployments-from-pull-requests-with-github-actions-13f890538e32)-like magic to sort that out

I'm just sorry I let you know at close of play on a Friday!

**Comment by arronlacey** [comment]
I did not mean to close that oops!

**Comment by malvikasharan** [comment]
Obviously I haven't yet managed to test this. @arronlacey can I assign this to you for review?

**Comment by sgibson91** [comment]
@malvikasharan Don't bother yet! The PR is broken, I need to do more work on it :)

**Comment by arronlacey** [comment]
> Obviously I haven't yet managed to test this. @arronlacey can I assign this to you for review?

sure thing @malvikasharan . When you've had a look @sgibson91 I'll review.

**Comment by da5nsy** [comment]
I'd be interested in helping out with this (I know little about CI but lots about image compression...)

**Comment by sgibson91** [comment]
@da5nsy thanks for the offer! Tbh, image compression isn't the issue right now, it's the fact that you cannot access GitHub Actions secrets in PRs made from forks. Which means that we'll have to do some CI/CD acrobatics like I describe in this blog post https://blog.jupyter.org/how-i-automated-authorised-cloud-deployments-from-pull-requests-with-github-actions-13f890538e32

File: .github/workflows/auto-compress-images.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      - '**.png'
+      - '**.gif'
+
+jobs:
+  compress-images:
+    runs-on: ubuntu-latest
+    environment: compress-images
+    permissions:
+      contents: write
+      pull-requests: write
```
Comment (sgibson91) [inline .github/workflows/auto-compress-images.yaml:18]:
```suggestion
    # Set permissions on GITHUB_TOKEN
    permissions:
      contents: write  # Make commits
      pull-requests: write  # Post a comment on a PR
```

**Comment by sgibson91** [comment]
OK, I managed to get this workflow into a state where I believe the secrets will work. This updates uses the `pull_request_target` trigger which runs the workflow in the context of the target repo with access to secrets. Generally speaking, it is not considered safe to combine the `pull_request_target` trigger with a checkout of the repo as this opens up vulnerability to code-injection attacks. However, one scenario described as "not vulnerable" to "checkout the code, lint it, and make a commit" and I believe this image compressions workflow falls into that category. We do not build the book or any code in this workflow.

The blog post I am using for reference:

* https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

@arronlacey when you're able to, please review!

**Comment by da5nsy** [comment]
I'm a bit concerned - it looks like crush pics only uses jpeg compression, which is distinctly suboptimal for the type of illustrations that make up most of the images in the book, since they have a lot of block colors (which can be well compressed by png index color compression, but look dreadful with jpeg compression).

## Here's a demo:

### Original image (and crop): [[source](https://github.com/alan-turing-institute/the-turing-way/blob/3997fa48f7532956975ade058b8376427c2c8a6c/book/website/figures/welcome.jpg)] (1,156kb)

welcome
welcome_zoom

### Crush.pics (balanced) (359kb)
(note the fuzzy edges)

welcome_crushpics
welcome_crushpics_zoom

### [Tiny png](https://tinypng.com/) (1056kb)

welcome_tinypng
welcome_tinypng_zoom

### Local png conversion (indexed color, Local (Perceptual), 12 colors) (210kb)

welcome_12
welcome_12_zoom

---
The example image I chose isn't actually perhaps the best example, since it is massive (4005px x 2703px) - the jpeg compression would be much more noticeable on smaller images.

**Comment by sgibson91** [comment]
@da5nsy can you recommend an alternative? I just went with the first action I found and I'm not strongly intent on using crushpics :)

**Comment by da5nsy** [comment]
I _think_ [this one](https://github.com/calibreapp/image-actions) might be a good bet.

I'm a little confused about how we'd set it up, but maybe that will make sense to you.

The bit I'm interested in is:
> [pngQuality](http://sharp.pixelplumbing.com/en/stable/api-output/#png): Number, integer 1-100, default 80 stored in a string.

The link there shows some options that would work nicely for our type of images:

- With `options.palette` set to `true` and `options.colours` set to something like `12`.
- OR `options.quality` set to something sensible (I'm not quite sure how this works... we'd need to play around with it).

**Comment by sgibson91** [comment]
@da5nsy Unfortunately, [these are the only inputs we can set](https://github.com/calibreapp/image-actions/blob/4570534b236f30f5df16fb699611bda9b895265d/action.yml#L5-L38). If we really wanted to set `options.palette`/`options.colours`/`options.quality`, it would require upstream changes to the action itself. Is there something we can do with the available inputs that would suffice?

**Comment by da5nsy** [comment]
Do you think `pngQuality` might get passed straight to `options.quality`? If so, that could definitely work.
How about we do some testing together sometime?

**Comment by sgibson91** [comment]
It's hard for me to say for certain because I don't read JavaScript, but I think it might get passed, via a config file, to something calling itself `quality`

https://github.com/calibreapp/image-actions/blob/4570534b236f30f5df16fb699611bda9b895265d/src/config.ts#L35-L42

I'll set it up on my fork of the repo using the default settings of the actions and we can go from there

**Comment by sgibson91** [comment]
Ok, PR showing with the compressed images is here: https://github.com/sgibson91/the-turing-way/pull/114

**Comment by da5nsy** [comment]
> Ok, PR showing with the compressed images is here: [sgibson91#114](https://github.com/sgibson91/the-turing-way/pull/114) Should be able to visually inspect the files in the Files Changed tab

Hi @sgibson91 - sorry it's taken me a while to get to this. It looks like it's not compressing the pngs using index-based compression. Looking at [this image](https://github.com/sgibson91/the-turing-way/blob/206477767fab991fcf05b02001542802e01f2329/book/website/_build/html/_images/reasons-reproducibility.png) (for example), I can see that the file attributes list it as having a bit depth of 32bits, whereas if I manually convert it to use a 7-color index it reads as 8-bit (I'm not sure why it doesn't read as 7...).
Screenshot_20220818_024139

As a result of this the image size drops from 117kb to 39kb, essentially losslessly (vs 174kb pre-compression)
Screenshot_20220818_024310
.

**Comment by sgibson91** [comment]
Thanks @da5nsy. I trust your input, but honestly it means nothing to me as I don't know much about image compression 😅 So I'm looking for guidance on what to do next?

**Comment by da5nsy** [comment]
I propose we shelve my ~~interest~~ obsession with indexed colour pngs and move forward as is. It would be nice to use, but I don't think it should block merging this as is. If we work it out later - great, if not - c'est la vie.

Separately - at some point I might go through existing images and do some gnarly compression at some point.

**Comment by sgibson91** [comment]
Thank you for providing your expertise on this @da5nsy!

**Review by da5nsy — APPROVED** [review]

**Comment by da5nsy** [comment]
I haven't tested this, but approve of the principle

**Pull request merged**